From f4c667722586955d310979e423febafa516797b8 Mon Sep 17 00:00:00 2001
From: Guillem Jover <guillem@hadrons.org>
Date: Wed, 2 Sep 2020 23:26:31 +0200
Subject: [PATCH 3/7] ping, ping6: Check setuid() return value and abort on
 failure

Even for root the setuid() call can fail, so we need to check whether it
has failed and abort the program at that point, as these are usually
installed as set-user-ID root programs.

Based-on-patch-by: Jayakrishna Vadayath <jvadayat@asu.edu>
---
 ping/ping.c  | 3 ++-
 ping/ping6.c | 3 ++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/ping/ping.c b/ping/ping.c
index eb5a7b06..33cf3858 100644
--- a/ping/ping.c
+++ b/ping/ping.c
@@ -294,7 +294,8 @@ main (int argc, char **argv)
   ping_set_sockopt (ping, SO_BROADCAST, (char *) &one, sizeof (one));
 
   /* Reset root privileges */
-  setuid (getuid ());
+  if (setuid (getuid ()) < 0)
+    error (0, errno, "setuid()");
 
   /* Force line buffering regardless of output device.  */
   setvbuf (stdout, NULL, _IOLBF, 0);
diff --git a/ping/ping6.c b/ping/ping6.c
index f10faf91..b9c42d7a 100644
--- a/ping/ping6.c
+++ b/ping/ping6.c
@@ -260,7 +260,8 @@ main (int argc, char **argv)
   setsockopt (ping->ping_fd, SOL_SOCKET, SO_BROADCAST, (char *) &one, sizeof (one));
 
   /* Reset root privileges */
-  setuid (getuid ());
+  if (setuid (getuid ()) < 0)
+    error (0, errno, "setuid()");
 
   /* Force line buffering regardless of output device.  */
   setvbuf (stdout, NULL, _IOLBF, 0);
-- 
2.30.0.284.gd98b1dd5eaa7

